### ISSUE TITLE
As an enterprise user, I want to change the columns to be shown in my catalogue, so that I can customise the page to my needs

### ISSUE THREAD

**Issue by mariocarabotta** [openfoodfoundation/openfoodnetwork#11055]

**Comment by NickIkenma** [comment]
I guess this is ready for dev since it seems to have been designed already.
Or do you want me to create an interactive prototype that shows all the scenarios, so it's easier to visualize and understand?

**Comment by mariocarabotta** [comment]
a similar inline button loading element has been implemented in the Connected Apps section in settings for Australia

https://staging.openfoodnetwork.org.au/admin/enterprises/anglesea-riverbank-market/edit#/connected_apps_panel

Let's talk about it during kickoff to avoid rebuilding too much

**Comment by mariocarabotta** [comment]
scenario 1 and 2 not considered for this estimate, will be developed separately (note in https://github.com/openfoodfoundation/openfoodnetwork/issues/11200)

**Comment by mariocarabotta** [comment]
@mariocarabotta to update, remove button and have autosave.
we'll create a separate card for adding a button if people ask for it

**Comment by dacook** [comment]
Just pointing out that we are currently considering an alternative to StimulusReflex, so the second part of this (scenarios 7,8,9) will need further discussion.

**Comment by dacook** [comment]
> @mariocarabotta to update, remove button and have autosave.

From memory, this is preferred because it's easier for the user, it's one less click. And it doesn't matter if they forget to save.
It could perhaps be annoying if they didn't want to save it, but it's a small list so quite easy to change back as needed.
It's probably a good time to do it while implementing the new design, people might be happier to accept the new behaviour.

But I think we might need a design for autosave. I can envisage two ways to design it:
1. Save on each click, and blur and show status as per current design. It _should_ be quick to save, but if it's not it would be quite annoying if you wanted to click multiple checkboxes, having to wait for it to save between each one. 
  - 1.1 perhaps we could avoid the blurring part, and allow the user to keep clicking. (this could fire off several requests in quick succession, but is probably fine)
  - It's probably fine either way, as long as they're implemented efficiently (with turbo streams or json layouts). 
2. Save when closing the popup (click or tab outside). But then we have to show status some other way

Alternatively we stick to current behaviour as designed and described above.
It's hard to say if any of these options would be quicker to build or not, but I suspect the current design would be quickest.

So, I think we can do option 1 without needing further design.
 I would implement it with a small turbo frame for the purpose of showing the status (saving/saved/failed), that way we don't need any custom JS.

**Comment by mariocarabotta** [comment]
I would not include any blurring, and I would consider starting without any saving feedback at all. Saving calls can succeed or fail in the background.
If we find this a bit too jarring / incomplete, we can add some loading / success / failure later.

What do you think?

**Comment by RachL** [comment]
@mariocarabotta @dacook I'm sorry, I feel this comment comes after the party (I missed the previous comment about autosave), but are we sure we want to introduce autosave as part of the first release? I understand it may saves us time when developing, but we need to consider testing and potential bugs during first release.
Given we are a bit late with BUU first launch, I'm a bit anxious to add something new now :/ But happy to discuss it.

**Comment by chahmedejaz** [comment]
Hi @dacook , @mariocarabotta - I found we are using multi-select dropdown in the reports as well and here's its design:
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/59338032/b096b736-5dfb-4b90-90b4-65060aa7632c)

However, as per the given mockup we want this design:
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/59338032/8f6bdff2-e9de-405e-b2bc-eccc14a54dd9)

Just wanted to confirm the path forward before moving forward:

1. Should we create the new multi-select dropdown as per the product's design?
2. Should reuse the existing component?
3. Should we update the existing report's dropdown and reuse it here?

Thanks.

**Comment by mariocarabotta** [comment]
I believe it would be good to implement the new design. the ones in report have uppercase (hard to read), some spacing issues, no colors. But also conscious of timeframes, @RachL it would be good to get your view as well. We can also go ahead with using the existing report dropdown to speed things up.

**Comment by dacook** [comment]
Thanks all. I suggest that we investigate if it's very easy to reuse the existing (Angular) component on this page.
If it is easy, let's do that for the first iteration.

But if it's not easy, we should go ahead with building a new component. (I think we can still re-use the `/admin/column_preferences/bulk_update` endpoint though, see usage in`app/assets/javascripts/admin/index_utils/services/columns.js.coffee`).

And so I don't think it's worth updating the report's dropdown yet. We can make the decision about that after this is done first.

**Comment by RachL** [comment]
Yes unfortunately we need to aim at the quickest solutions now for BUU. Not only in terms of timeframes, but also in terms of budget (yes I know that rebuilding means more budget in the end, but that's what our cashflow situation forces us to do).

**Comment by mariocarabotta** [comment]
ok that sounds reasonable. 
now @chahmedejaz is on [#12398](https://github.com/openfoodfoundation/openfoodnetwork/issues/12398).
after that, we already agreed he'll investigate if it's easy enough to reuse the report dropdown.

thanks!

**Comment by dacook** [comment]
I'm going to try the spike now to see which path to take.

**Comment by dacook** [comment]
Ok, I got a proof of concept going. It's actually not too bad, although there's a slight delay after page load before the columns are shown/hidden.

Remaining:
1. need to update a couple of colspans in _table partial with angular. I don't know how to do that yet.
2. re-init after turbo frame load (eg after save changes)
2. move dropdown to desired position (to the right of page size dropdown)
2. Create a new controller instead of breaking the current AdminProductEditCtrl
3. Optional: re-style as per design

https://github.com/dacook/openfoodnetwork/commit/1351b793ea5999d12d7582a26c411cadf7a748d5
![Screenshot 2024-05-29 at 1 59 04 pm](https://github.com/openfoodfoundation/openfoodnetwork/assets/4188088/ac90a37d-e1af-4ae7-ba00-2a69dac47215)


Given it's pretty close, I think it's worth proceeding. I'll timebox another 30mins to see if I can do items 1-2 to further prove the concept.

**Comment by dacook** [comment]
Question: Do we want to keep user column preferences from the old screen and bring them to the new screen?

So far, I've built it so that the preferences from each screen are separate. 

I think it won't be simple to share the column prefs between both screens because there are some differences (different order, and some different columns). I can see a couple of ways to do it:
* add a layer in column_preferences_controller to load preferences from the old screen, if no other preferences set yet
* a one-off migration to copy column preferences from old screen to new.

Hmm.. I think it's not worth it, so will make the call and won't raise it further. That is, column preferences from the old screen will not be brought to the new screen.

**Comment by mariocarabotta** [comment]
I'd say not worth it too..it's a few clicks away for people, not a big deal
